### ISSUE TITLE
Current editor behaviour promotes extraneous trailing whitespace

### ISSUE THREAD

**Issue by axefrog** [Microsoft/vscode#15923]
- VSCode Version: 1.7.2
- OS Version: Windows 10

Steps to Reproduce:

1. Do anything that would cause trailing whitespace, such as deleting a newline when the following line contains only whitespace, or typing some characters, then some whitespace in preparation for the next word, but then changing your mind and moving the cursor elsewhere.
2. Go to another line, or press HOME to go back to the start of the current line.
3. With the cursor anywhere before the end of the original line, press END to jump back to the end, and start typing again.

**The END key jumps to the last non-whitespace character instead of the last character**. This makes sense for the HOME key, because you usually want to go to the start of the current line indentation, but the END key should not have this behaviour because (a) if you typed trailing whitespace deliberately, you probably want to keep it, and (b) if you did not type trailing whitespace deliberately, you probably want to get rid of it, which requires first knowing it's there. Yes, you can turn on hidden whitespace, but this doesn't change things, it just makes this friction more visible while editing.

Because the END key tricks you into thinking that the line has no trailing whitespace, you end up with a lot of lines with one, two or three trailing spaces as a result of making small edits to code. You then go to commit code to the repo only to realise there is a bunch of extraneous whitespace you didn't previously notice. This is also an annoyance when jumping around the code in other ways, as the previously-hidden whitespace gets in the way of edits and needs to be corrected.

I actually made [an extension](https://marketplace.visualstudio.com/items?itemName=NathanRidley.autotrim) just to solve this and it currently has over 500 installs despite no promotion on my part, suggesting that others find this annoying too, and looked for an extension to solve the problem.

**Comment by alexandrudima** [comment]
In the core we have already two knobs for getting rid of extraneous whitespace:

### Trim trailing whitespace on save
Under the setting `files.trimTrailingWhitespace` (it is off by default). All trim whitespace gets removed when the file is saved explicitly (`ctrl+s`). If auto-save is on, the lines where a cursor resides **on** trailing whitespace will not be trimmed.

### Trim automatic whitespace on type
Under the setting `editor.trimAutoWhitespace` (it is on by default). This removes automatically inserted whitespace by the editor, such as that that gets inserted when pressing Enter after an indented line and then changing ones mind and typing somewhere else or pressing Enter repeatedly.

I take the suggestion to reconsider the behaviour of <kbd>End</kbd>. Perhaps it would be a very good idea to make it jump first to after the trailing whitespace, and then jump to before the trailing whitespace (sort of opposite to <kbd>Home</kbd>).

**Comment by axefrog** [comment]
> Trim trailing whitespace on save
> Under the setting files.trimTrailingWhitespace (it is off by default). All trim whitespace gets removed when the file is saved explicitly (ctrl+s). If auto-save is on, the lines where a cursor resides on trailing whitespace will not be trimmed.

The main friction I experience is *during* editing. It's good that the extraneous whitespace can be removed when saving, but the annoyance is when you make an edit that should result in a single space between tokens and the additional unseen whitespace gets in the way and needs to be edited.

> Trim automatic whitespace on type
> Under the setting editor.trimAutoWhitespace (it is on by default). This removes automatically inserted whitespace by the editor, such as that that gets inserted when pressing Enter after an indented line and then changing ones mind and typing somewhere else or pressing Enter repeatedly.

What I found was that I didn't want the editor trying to second-guess my whitespace choices. I didn't need it to be trying to neaten things up for me according to a general set of rules about which whitespace to keep and which to remove, but rather I needed to not be accidentally appending small chunks of whitespace to the ends of lines as a result of the quirk of the editor, and that's what I set out to fix with my extension. It works like this:

- The current line is never modified while you are editing it
- Any lines with active cursors are kept track of
- When a line that previously had a cursor no longer has a cursor, the line is trimmed
- No edits are performed against lines that don't need trimming

If I were to make one addition to the extension, it would be to disqualify lines from being trimmed unless an edit has also been made to that line, thus preventing a file from being marked as unsaved when skimming through it with the cursor, when the file was only opened to quickly view some part of it, but not to make any edits.

> I take the suggestion to reconsider the behaviour of End. Perhaps it would be a very good idea to make it jump first to after the trailing whitespace, and then jump to before the trailing whitespace (sort of opposite to Home).

As you say, the behaviour of the End key I think is the core of the issue. If it went to the *actual* end of the line, then VS Code would behave the same as other editors and the treatment of the symptoms would no longer be necessary..
